### PR TITLE
DOC/API: Styler documentation changes

### PR DESCRIPTION
**Pull request by TomAugspurger**
Closes https://github.com/pydata/pandas/issues/13222, #13225
- Update docstring and notebook on output shapes matching
- Change build process to use nbconvert to execute the notebook and then convert using NBConvert.

Second commit puts some more restrictions on the output of a user's function passed to `.apply`.

### CODE REVIEW

File: pandas/formats/style.py [hunk 1]
```diff
@@ -444,15 +444,17 @@ def apply(self, func, axis=0, subset=None, **kwargs):
...
+        func : function
+            ``func`` should take a Series or DataFrame (depending
+            on ``axis``), and return an object with the same shape
+        axis : int, str or None
             apply to each column (``axis=0`` or ``'index'``)
             or to each row (``axis=1`` or ``'columns'``) or
-            to the entire DataFrame at once with ``axis=None``.
-        subset: IndexSlice
+            to the entire DataFrame at once with ``axis=None``
+        subset : IndexSlice
```
Comment (jreback) [inline pandas/formats/style.py]: I thought we don't have a space before the :

? in general I mean

File: pandas/formats/style.py [hunk 2]
```diff
@@ -461,9 +463,22 @@ def apply(self, func, axis=0, subset=None, **kwargs):
 
         Notes
         -----
+        The output shape of ``func`` should match the input, i.e. if
+        ``x`` is the input row, column, or table (depending on ``axis``),
+        then ``func(x.shape) == x.shape`` should be true.
+
```
Comment (jreback) [inline pandas/formats/style.py]: this is very easy to assert / check

you should just enforce it

File: pandas/formats/style.py (hunk 1 above)
Comment (TomAugspurger) [inline pandas/formats/style.py]: The examples here do: https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt#sections, but I don't think it's enforced, does pandas have a preference?

Comment (jreback) [inline pandas/formats/style.py]: not sure we are consistent as have seen both ways (or does it even format differently)?

Comment (jreback) [inline pandas/formats/style.py]: I am wrong - it does specify spaces on both sides
(contrast this to linting which doesn't have s space - but different issue)

File: doc/make.py [hunk 3]
```diff
@@ -102,10 +105,88 @@ def clean():
...
+    import nbformat
+    from nbconvert.preprocessors import ExecutePreprocessor
+
+    with io.open(src, encoding='utf-8') as f:
+        nb = nbformat.read(f, as_version=4)
+
+    ep = ExecutePreprocessor(allow_errors=allow_errors,
+                             timeout=timeout,
+                             kernel_name=kernel_name)
+    ep.preprocess(nb, {'metadta': {'path': 'notebooks/'}})
```
Comment (jorisvandenbossche) [inline doc/make.py]: metadta -> metadata (side question: what does this do?)

File: pandas/formats/style.py (hunk 1 above)
Comment (jorisvandenbossche) [inline pandas/formats/style.py]: yes, it definitly specify spaces on both sides, it's a comment I often give when reviewing docstrings ... :-) (and it is needed to get the html formatting correctly using numpydoc, don't know if napoleon is more flexible)

File: pandas/formats/style.py [hunk 4]
```diff
@@ -444,15 +444,17 @@ def apply(self, func, axis=0, subset=None, **kwargs):
 
         Parameters
         ----------
-        func: function
-        axis: int, str or None
+        func : function
+            ``func`` should take a Series or DataFrame (depending
```
Comment (jorisvandenbossche) [inline pandas/formats/style.py]: And if we are doing nitpicky :-) The numpydoc standard says that references to keywords should have single backticks ...

File: pandas/formats/style.py (hunk 1 above)
Comment (jreback) [inline pandas/formats/style.py]: maybe we should have a commit hook to check this? or lint check or something?

Comment (jreback) [inline pandas/formats/style.py]: https://gitlab.com/pycqa/flake8-docstrings anyone try?

**Comment by jorisvandenbossche** [comment]
@TomAugspurger Did you change a lot in the notebook itself? (Or other phrased: is it worth reviewing? Because online the diff is not shown)

File: pandas/formats/style.py (hunk 1 above)
Comment (jorisvandenbossche) [inline pandas/formats/style.py]: Last time I looked for something like this, it only checked for pep257, and not for numpydoc specific things

Comment (jorisvandenbossche) [inline pandas/formats/style.py]: But there seems some discussion/work under way: https://github.com/PyCQA/pydocstyle/issues/129

Comment (jreback) [inline pandas/formats/style.py]: yes that looks good / we should follow that release

**Comment by TomAugspurger** [comment]
@jorisvandenbossche the notebook changes were

Added:

>For `Styler.applymap` your function should take a scalar and return a single string with the CSS attribute-value pair.
For `Styler.apply` your function should take a Series or DataFrame (depending on the axis parameter), and return a Series or DataFrame with an identical shape where each value is a string with a CSS attribute-value pair.

and

> And crucially the input and output shapes of `func` must match. If `x` is the input then ``func(x).shape == x.shape``.

The  reason the diff is so big is that I deleted all the output from the notebook committed to the repo.

File: pandas/formats/style.py (hunk 4 above)
Comment (TomAugspurger) [inline pandas/formats/style.py]: Ah, I thought we decided to use double ticks since it looks nicer in our HTML :)

File: doc/make.py (hunk 3 above)
Comment (TomAugspurger) [inline doc/make.py]:
```
resources : dictionary
    Additional resources used in the conversion process. For example,
    passing ``{'metadata': {'path': run_path}}`` sets the
    execution path to ``run_path``.
```

But apparently nbconvert doesn't do any argument checking :) I'll remove that.

File: pandas/formats/style.py (hunk 2 above)
Comment (TomAugspurger) [inline pandas/formats/style.py]: I'll push commit tonight with this (very easy, two line change). I had written a test that violated the rule, but "worked" 😳 . Just need to write additional tests.

**Comment by jreback** [comment]
@TomAugspurger what about have a shape check when using ``.apply``? e.g. to guarntee (or otherwise raise)?

**Comment by TomAugspurger** [comment]
> @TomAugspurger what about have a shape check when using .apply

I'll push that change tonight, it's pretty easy, just need tests.

**Comment by jreback** [comment]
gr8! @TomAugspurger better to have an exception than documentation. Further no issues with restricting ``.apply``. No need to make this too general (like ``DataFrame.apply``) that just leads to blind pathways.

**Comment by TomAugspurger** [comment]
Sorry for the delay, pushed changes.

1. The result of `Styler.apply(func, axis=0/1)` must have the same shape. The output is from `DataFrame.apply`, so we know we have a Series with the correct index labels.
2. The result of `Styler.apply(func, axis=None)` must be a DataFrame with the same index columns / labels (and so same shape as well).

I put the second one in place to avoid confusing output like

screen shot 2016-05-20 at 7 14 39 am

Only the top row gets colored, even though the output was 2x2.

**Comment by jreback** [comment]
xref: https://travis-ci.org/pydata/pandas/jobs/132906479#L1631

**Comment by TomAugspurger** [comment]
Sorry, forgot this was still open.

Here are the relevant changes to the notebook

>  Notice that the output shape of highlight_max matches the input shape, an array with len(s) items.

and

> When using ``Styler.apply(func, axis=None)``, the function must return a DataFrame with the same index and column labels.

and

> And crucially the input and output shapes of `func` must match. If `x` is the input then ``func(x).shape == x.shape``.

File: doc/README.rst [hunk 5]
```diff
@@ -81,7 +81,7 @@ have ``sphinx`` and ``ipython`` installed. `numpydoc
 <https://github.com/numpy/numpydoc>`_ is used to parse the docstrings that
 follow the Numpy Docstring Standard (see above), but you don't need to install
 this because a local copy of ``numpydoc`` is included in the pandas source
-code.
+code. To build the Jupyter notebooks included in the documentation, you'll need `nbconvert <https://nbconvert.readthedocs.io/en/latest/>`_ and `nbformat <http://nbformat.readthedocs.io/en/latest/>`_.
```
Comment (jreback) [inline doc/README.rst]: we might also have this same documenation in ``contributing.rst`` so update there as well (as well as some info on what deps you need to build the docs)

File: doc/make.py [hunk 6]
```diff
@@ -102,10 +105,90 @@ def clean():
...
+        yield
+    finally:
+        try:
+            os.remove(nb + '.executed')
+        except OSError:
+            pass
+
+
+def execute_nb(src, dst, allow_errors=False, timeout=1000, kernel_name=''):
+    '''
```
Comment (jreback) [inline doc/make.py]: use """

File: doc/make.py [hunk 7]
```diff
@@ -102,10 +105,90 @@ def clean():
...
+                             kernel_name=kernel_name)
+    ep.preprocess(nb, resources={})
+
+    with io.open(dst, 'wt', encoding='utf-8') as f:
+        nbformat.write(nb, f)
+    return dst
+
+
+def convert_nb(src, dst, to='html', template_file='basic'):
+    '''
```
Comment (jreback) [inline doc/make.py]: same

File: doc/source/whatsnew/v0.18.2.txt [hunk 8]
```diff
@@ -141,6 +141,7 @@ API changes
 - Non-convertible dates in an excel date column will be returned without conversion and the column will be ``object`` dtype, rather than raising an exception  (:issue:`10001`)
 - An ``UnsupportedFunctionCall`` error is now raised if NumPy ufuncs like ``np.mean`` are called on groupby or resample objects (:issue:`12811`)
 - Calls to ``.sample()`` will respect the random seed set via ``numpy.random.seed(n)`` (:issue:`13161`)
+- ``Styler.apply`` is now stricter about the outputs your function must return. For ``axis=0`` or ``axis=1``, the output shape must be identical. For ``axis=None``, the output must be a DataFrame with identical columns and index labels. (:issue:`13222`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.18.2.txt]: more strict

File: pandas/formats/style.py [hunk 9]
```diff
@@ -444,15 +462,19 @@ def apply(self, func, axis=0, subset=None, **kwargs):
 
         Parameters
         ----------
-        func: function
-        axis: int, str or None
+        func : function
```
Comment (jreback) [inline pandas/formats/style.py]: you can change parameters to ``func:`` (no space)

File: doc/README.rst (hunk 5 above)
Comment (jorisvandenbossche) [inline doc/README.rst]: Maybe we can just include this file in contributing.rst, to not duplicate things (I thought it was done like that before)

File: pandas/formats/style.py [hunk 10]
```diff
@@ -427,11 +427,29 @@ def _compute(self):
...
         if axis is not None:
-            result = self.data.loc[subset].apply(func, axis=axis, **kwargs)
+            result = data.apply(func, axis=axis, **kwargs)
         else:
-            # like tee
-            result = func(self.data.loc[subset], **kwargs)
+            result = func(data, **kwargs)
+            if not isinstance(result, pd.DataFrame):
+                raise TypeError("Function {!r} must return a DataFrame when "
+                                "passed to `Styler.apply` with axis=None")
```
Comment (jorisvandenbossche) [inline pandas/formats/style.py]: Isn't there a `.format(..)` missing? (given the `{!r}`)

File: pandas/formats/style.py [hunk 11]
```diff
@@ -427,11 +427,29 @@ def _compute(self):
...
+            if not (result.index.equals(data.index) and
+                    result.columns.equals(data.columns)):
+                msg = ('Result of {!r} must have identical index and columns '
+                       'as the input')
```
Comment (jorisvandenbossche) [inline pandas/formats/style.py]: same here

**Comment by jorisvandenbossche** [comment]
@TomAugspurger Does it build locally for you? (the docs) As on travis there is an error "InputError: [Errno 2] No such file or directory: 'source/html-styling.html'" (not sure if that means that something went wrong ..)

**Comment by TomAugspurger** [comment]
Addressed all the comments, thanks. @jorisvandenbossche it does build locally, running from `pandas/doc`. Is that failure from before this pull request? I'll check this output of this Travis run.

**Comment by jorisvandenbossche** [comment]
Let's see how the travis docs will look like after it is merged.

**Pull request closed without merging**